### PR TITLE
cgmgr: reuse dbus connection

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind bug
#### What this PR does / why we need it:
reuse dbus connection and attempt to reconnect if we get a dbus error
this PR heavily adapts https://github.com/opencontainers/runc/blob/7362fa2d282feffb9b19911150e01e390a23899d/libcontainer/cgroups/systemd

eventually, the plan is to upstream the internal/dbusmgr package to godbus, but we're picking this here first
#### Does this PR introduce a user-facing change?

```release-note
Reuse connection when connecting to dbus, as well as reattempt the connection if it fails temporarily
```

### CODE REVIEW

File: utils/utils.go [hunk 1]
```diff
@@ -70,18 +72,22 @@ func RunUnderSystemdScope(pid int, slice, unitName string, properties ...systemd
...
+		})
+
 		if err == nil {
 			break
 		}
 		if !errors.Is(err, syscall.EAGAIN) {
 			return err
 		}
 	}
-	defer conn.Close()
```
Comment (haircommander) [inline utils/utils.go:81]: it's not really clear to me how we call close on this anymore

Comment (haircommander) [reply utils/utils.go:81]: wait we're reusing this connection why would we close it :facepalm:

File: internal/dbusmgr/dbusmgr.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	dbusRootless bool
+)
+
+type DbusConnManager struct {
+}
+
+// NewDbusConnManager initializes systemd dbus connection manager.
+func NewDbusConnManager(rootless bool) *DbusConnManager {
+	if dbusInited && rootless != dbusRootless {
+		panic("can't have both root and rootless dbus")
```
Comment (kolyshkin) [inline internal/dbusmgr/dbusmgr.go:30]: Nice, I was thinking about adding something like this but never got to it, mostly because it's not possible in a short-lived binary like `runc` (which is either rootless or not, can't be both).

File: internal/dbusmgr/dbusmgr.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+var (
+	dbusC        *systemdDbus.Conn
+	dbusMu       sync.RWMutex
+	dbusInited   bool
+	dbusRootless bool
+)
+
+type DbusConnManager struct {
+}
```
Comment (kolyshkin) [inline internal/dbusmgr/dbusmgr.go:25]: I guess it needs to be
```go
type DbusConnManager struct {}
```

(from the `gofumpt` POV)

File: internal/dbusmgr/dbusmgr.go (hunk 2 above)
Comment (kolyshkin) [reply internal/dbusmgr/dbusmgr.go:30]: But you do need to
(1) take a mutex before getting to these variables, otherwise races are possible (if two goroutines call `NewDbusConnManager`)
(2) do set `dbusInited = true` -- otherwise this check never works.

Comment (haircommander) [reply internal/dbusmgr/dbusmgr.go:30]: hahaha then you can thank your past-self https://github.com/opencontainers/runc/blame/2f8e8e9d9775003bf76730a8a1c10f29b0000edf/libcontainer/cgroups/systemd/dbus.go#L25

File: utils/utils.go [hunk 4]
```diff
@@ -70,18 +72,22 @@ func RunUnderSystemdScope(pid int, slice, unitName string, properties ...systemd
...
 
 	// Block until job is started
 	<-ch
+	close(ch)
```
Comment (kolyshkin) [inline utils/utils.go:90]: This belongs to a separate commit I think.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
LGTM aside from the hunk which do not belong to the commit.

Given my history with this feature, I think it might make sense to at least manually test it.

The test would be to start/stop lots of containers, doing `while :; do systemctl restart dbus.service; done` in parallel. Before this commit, this should inevitably result in errors from container start (i.e. `RunUnderSystemdScope`). This needs to be checked, too, that the test is valid.

Once test validity is established, test the patched version to make sure the error is no more.

File: internal/dbusmgr/dbusmgr.go (hunk 2 above)
Comment (kolyshkin) [reply internal/dbusmgr/dbusmgr.go:31]: 🤦🏻

Comment (kolyshkin) [reply internal/dbusmgr/dbusmgr.go:31]: https://github.com/opencontainers/runc/pull/3006

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by mrunalp — APPROVED** [review]
/lgtm
